core/src/main/scala/Diet.scala

```diff
@@ -0,0 +1,321 @@
+/**
+ * Created by Nicolas A Perez (@anicolaspp) on 2/4/16.
+ */
+
+package dogs
+
+import dogs.Order._
+import dogs.Predef._
+import dogs._
+
+import scala.annotation.tailrec
+
+
+
+/**
+ * Represent discrete operations that can be performed on A
+ */
+trait Discrete[A] extends Order[A] {
+
+  /**
+   * return the successor of x
+   */
+  def succ(x: A): A
+
+  /**
+   * return the predecessor of
+   */
+  def pred(x: A): A
+
+  /**
+   * verify if x and y are consecutive
+   */
+  def adj(x: A, y: A): Boolean = succ(x) == y
+}
+
+/**
+ * Represent a range [x, y] that can be generated using generate() or reverse().
+ */
+trait ARange[A] {
```

Comment:
I think is should be part of the public interface. At this point we are not adding an ARange, but it can be implemented at some point. This comment also responds to one other other comments